prow/plank/controller.go

```diff
@@ -0,0 +1,275 @@
+		updated = true
+		pj.Status.CompletionTime = time.Now()
+		pj.Status.State = kube.FailureState
+		if err := c.report(pj); err != nil {
+			return fmt.Errorf("error reporting to crier: %v", err)
+		}
+	} else {
+		// Pod is running. Do nothing.
+	}
+	if updated {
+		_, err := c.kc.ReplaceProwJob(pj.Metadata.Name, pj)
+		return err
+	} else {
+		return nil
+	}
+}
+
+func (c *Controller) report(pj kube.ProwJob) error {
+	if !pj.Spec.Report {
+		return nil
+	}
+	if len(pj.Spec.Refs.Pulls) != 1 {
+		return fmt.Errorf("prowjob %s has %d pulls, not 1", pj.Metadata.Name, len(pj.Spec.Refs.Pulls))
+	}
+	return crier.ReportToCrier(c.crierURL, crier.Report{
+		RepoOwner:    pj.Spec.Refs.Org,
+		RepoName:     pj.Spec.Refs.Repo,
+		Author:       pj.Spec.Refs.Pulls[0].Author,
+		Number:       pj.Spec.Refs.Pulls[0].Number,
+		Commit:       pj.Spec.Refs.Pulls[0].SHA,
+		Context:      pj.Spec.Context,
+		State:        string(pj.Status.State),
+		Description:  pj.Spec.Description,
+		RerunCommand: pj.Spec.RerunCommand,
+		URL:          pj.Spec.URL,
+	})
+}
+
+func (c *Controller) startPod(pj kube.ProwJob) (string, error) {
+	buildID, err := c.getBuildID(c.totURL, pj.Spec.Job)
+	if err != nil {
+		return "", fmt.Errorf("error getting build ID: %v", err)
+	}
+	spec := pj.Spec.PodSpec
+	spec.RestartPolicy = "Never"
+	spec.NodeSelector = map[string]string{
+		"role": "build",
+	}
+	// keep this synchronized with get_running_build_log in Gubernator!
+	podName := fmt.Sprintf("%s-%s", pj.Spec.Job, buildID)
+	if len(podName) > 60 {
+		podName = podName[len(podName)-60:]
+	}
+
+	for i := range spec.Containers {
+		spec.Containers[i].Name = fmt.Sprintf("%s-%d", podName, i)
+		spec.Containers[i].Env = append(spec.Containers[i].Env,
+			kube.EnvVar{
+				Name:  "JOB_NAME",
+				Value: pj.Spec.Job,
+			},
+			kube.EnvVar{
+				Name:  "BUILD_NUMBER",
+				Value: buildID,
+			},
+		)
+		if pj.Spec.Type == kube.PeriodicJob {
+			continue
+		}
+		spec.Containers[i].Env = append(spec.Containers[i].Env,
+			kube.EnvVar{
+				Name:  "REPO_OWNER",
+				Value: pj.Spec.Refs.Org,
+			},
+			kube.EnvVar{
+				Name:  "REPO_NAME",
+				Value: pj.Spec.Refs.Repo,
+			},
+			kube.EnvVar{
+				Name:  "PULL_BASE_REF",
+				Value: pj.Spec.Refs.BaseRef,
+			},
+			kube.EnvVar{
+				Name:  "PULL_BASE_SHA",
+				Value: pj.Spec.Refs.BaseSHA,
+			},
+			kube.EnvVar{
+				Name:  "PULL_REFS",
+				Value: pj.Spec.Refs.String(),
+			},
+		)
+		if pj.Spec.Type == kube.PostsubmitJob || pj.Spec.Type == kube.BatchJob {
+			continue
+		}
+		spec.Containers[i].Env = append(spec.Containers[i].Env,
+			kube.EnvVar{
+				Name:  "PULL_NUMBER",
+				Value: strconv.Itoa(pj.Spec.Refs.Pulls[0].Number),
+			},
+			kube.EnvVar{
+				Name:  "PULL_PULL_SHA",
+				Value: pj.Spec.Refs.Pulls[0].SHA,
+			},
+		)
+		// Set the HostPort to 9999 for all build pods so that they are forced
+		// onto different nodes. Once pod affinity is GA, use that instead.
+		spec.Containers[i].Ports = append(spec.Containers[i].Ports,
```

Comment:
It's just to make sure that only one pod runs per node. We will probably end up removing this at some point. The number itself is meaningless.